src/replication.c

```diff
@@ -1030,28 +1267,81 @@ read_io_resp(spec_t *spec, replica_t *replica, read_event_t *revent)
 	return donecount;
 }
 
+/*
+ * write data on management fd until EAGAIN
+ */
+int
+handle_write_data_event(replica_t *replica)
+{
+	io_event_t wevent;
+	int rc = -1;
+	mgmt_cmd_t *mgmt_cmd;
+
+	MTX_LOCK(&replica->r_mtx);
+	mgmt_cmd = TAILQ_FIRST(&replica->mgmt_cmd_queue);
+	if (!(mgmt_cmd != NULL &&
+		(mgmt_cmd->mgmt_cmd_state == WRITE_IO_SEND_HDR ||
+		mgmt_cmd->mgmt_cmd_state == WRITE_IO_SEND_DATA))) {
+		MTX_UNLOCK(&replica->r_mtx);
+		return rc;
+	}
+
+	MTX_UNLOCK(&replica->r_mtx);
+
+	wevent.fd = replica->mgmt_fd;
+	wevent.state = &(mgmt_cmd->mgmt_cmd_state);
+	wevent.io_hdr = mgmt_cmd->io_hdr;
+	wevent.io_data = (void **)(&(mgmt_cmd->data));
+	wevent.byte_count = &(mgmt_cmd->io_bytes);
+
+	rc = write_io_data(replica, &wevent);
+	return rc;
+}
+
 /*
  * reads data on management fd until EAGAIN
  */
 void
-handle_read_data_event(int fd)
+handle_read_data_event(replica_t *replica)
 {
-	spec_t *spec = NULL;
-	read_event_t revent;
+	io_event_t revent;
+	mgmt_cmd_t *mgmt_cmd;
 
-	replica_t *replica = get_replica(fd, &spec);
-	if (replica == NULL || spec == NULL) {
-		REPLICA_ERRLOG("error in getting replica %p/spec %p for fd %d..\n", replica, spec, fd);
+	MTX_LOCK(&replica->r_mtx);
+	mgmt_cmd = TAILQ_FIRST(&replica->mgmt_cmd_queue);
+	if (!(mgmt_cmd != NULL &&
+		(mgmt_cmd->mgmt_cmd_state == READ_IO_RESP_HDR ||
+		mgmt_cmd->mgmt_cmd_state == READ_IO_RESP_DATA))) {
+		MTX_UNLOCK(&replica->r_mtx);
 		return;
 	}
 
-	revent.fd = fd;
-	revent.state = &(replica->mgmt_io_state);
-	revent.resp_hdr = replica->mgmt_ack;
-	revent.resp_data = (void **)(&(replica->mgmt_ack_data));
-	revent.read_count = &(replica->mgmt_io_read);
+	MTX_UNLOCK(&replica->r_mtx);
+	int count = 0;
+
+	revent.fd = replica->mgmt_fd;
+	revent.state = &(mgmt_cmd->mgmt_cmd_state);
```

Comment:
we can probably use mgmt_cmd_s itself rather than copying it to revent structure..